### PR TITLE
Add CAST(double as decimal)

### PR DESCRIPTION
**Pull request by wuxueyang96**

### CODE REVIEW

File: velox/expression/tests/CastExprTest.cpp [hunk 1]
```diff
@@ -173,6 +173,32 @@ class CastExprTest : public functions::test::CastBaseTest {
...
+        input,
+        makeFlatVector<int128_t>(
+            {-33'330'300'000'000,
+             -22'220'200'000'000,
+             -10'000'000'000,
+             0,
+             1'000'000'000'000,
+             999'999'900'000'000},
+            DECIMAL(20, 10)));
+  }
```
Comment (jinchengchenghh) [inline velox/expression/tests/CastExprTest.cpp:200]: Can you add the test to cast 99999.99 to decimal(20, 1) to test round up?

File: velox/expression/CastExpr.cpp [hunk 2]
```diff
@@ -241,6 +241,45 @@ void applyIntToDecimalCastKernel(
...
+    exec::EvalCtx& context,
+    const TypePtr& toType,
+    VectorPtr& castResult) {
+  auto sourceVector = input.as<SimpleVector<TInput>>();
+  auto castResultRawBuffer =
+      castResult->asUnchecked<FlatVector<TOutput>>()->mutableRawValues();
+  const auto& toPrecisionScale = getDecimalPrecisionScale(*toType);
+  context.applyToSelectedNoThrow(rows, [&](vector_size_t row) {
+    const auto& sourceValue = sourceVector->valueAt(row);
+    if (!std::isfinite(sourceValue)) {
```
Comment (jinchengchenghh) [inline velox/expression/CastExpr.cpp:260]: sourceVector->isNullAt(row)

File: velox/expression/CastExpr.cpp [hunk 3]
```diff
@@ -241,6 +241,45 @@ void applyIntToDecimalCastKernel(
...
+        static_cast<TInput>(DecimalUtil::kPowersOfTen[toPrecisionScale.second]);
+
+    if (toValue >= 0) {
+      toValue = std::floor(toValue + 0.5);
+    } else {
+      toValue = std::ceil(toValue - 0.5);
+    }
+
+    if (toValue <= static_cast<TInput>(std::numeric_limits<TOutput>::min()) ||
+        toValue >= static_cast<TInput>(std::numeric_limits<TOutput>::max())) {
```
Comment (jinchengchenghh) [inline velox/expression/CastExpr.cpp:275]: If value beyond the  range, it will throw exception in Presto.
```
presto:default> select cast((cast ('999999999999999999999.0' as double)) as decimal(18, 2));
Query 20230724_153614_00011_c9bdk failed: Cannot cast DOUBLE '1.0E21' to DECIMAL(18, 2)

presto:default> select cast((cast ('999999999999999999999.0' as double)) as decimal(38, 2));
           _col0
---------------------------
 1000000000000000000000.00
(1 row)
```

File: velox/expression/CastExpr.cpp [hunk 4]
```diff
@@ -241,6 +241,45 @@ void applyIntToDecimalCastKernel(
...
+    auto toValue = sourceValue *
+        static_cast<TInput>(DecimalUtil::kPowersOfTen[toPrecisionScale.second]);
+
+    if (toValue >= 0) {
+      toValue = std::floor(toValue + 0.5);
+    } else {
+      toValue = std::ceil(toValue - 0.5);
+    }
+
+    if (toValue <= static_cast<TInput>(std::numeric_limits<TOutput>::min()) ||
```
Comment (jinchengchenghh) [inline velox/expression/CastExpr.cpp:274]: And you should check the range by `DecimalUtil`
```
  FOLLY_ALWAYS_INLINE static void valueInRange(int128_t value) {
    VELOX_CHECK(
        (value >= kLongDecimalMin && value <= kLongDecimalMax),
        "Value '{}' is not in the range of Decimal Type",
        value);
  }
```

File: velox/expression/tests/CastExprTest.cpp [hunk 5]
```diff
@@ -173,6 +173,37 @@ class CastExprTest : public functions::test::CastBaseTest {
...
+        makeFlatVector<T>({-3333.03, -2222.02, -1.0, 0.00, 100, 99999.99});
+    testComplexCast(
+        "c0",
+        input,
+        makeFlatVector<int64_t>(
+            {-33'330'300, -22'220'200, -10'000, 0, 1'000'000, 999'999'900},
+            DECIMAL(10, 4)));
+
+    // floating to long decimal
+    testComplexCast(
```
Comment (jinchengchenghh) [inline velox/expression/tests/CastExprTest.cpp:189]: Please also add the test for the corner case and exception.
For example, test value > int64_max and <int64_max when output type is long decimal.

File: velox/expression/CastExpr.cpp (hunk 4 above)
Comment (wuxueyang96) [reply velox/expression/CastExpr.cpp:274]: VELOX_CHECK will throw an exception if the result of the expression is false. So that means, we should throw an exception here instead of setting a null value?

Comment (jinchengchenghh) [reply velox/expression/CastExpr.cpp:274]: Yes, like this exception https://github.com/facebookincubator/velox/blob/main/velox/type/DecimalUtil.h#L141

In Spark, it will set null instead of throwing exception, but now the cast behavior consists with Presto

File: velox/expression/CastExpr.cpp [hunk 6]
```diff
@@ -241,6 +241,45 @@ void applyIntToDecimalCastKernel(
...
+    if (!std::isfinite(sourceValue)) {
+      castResult->setNull(row, true);
+      return;
+    }
+
+    auto toValue = sourceValue *
+        static_cast<TInput>(DecimalUtil::kPowersOfTen[toPrecisionScale.second]);
+
+    if (toValue >= 0) {
+      toValue = std::floor(toValue + 0.5);
```
Comment (jinchengchenghh) [inline velox/expression/CastExpr.cpp:269]: If the toValue is Double::max - 0.4, It will overflow and cannot get correct result.

Comment (jinchengchenghh) [reply velox/expression/CastExpr.cpp:269]: Postgresql convert double to varchar and then convert to numeric type, https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/numeric.c#L4515

I have a PR to cast varchar to decimal https://github.com/facebookincubator/velox/pull/5307

Comment (jinchengchenghh) [reply velox/expression/CastExpr.cpp:269]: Here maybe cast double to int128, you can check and use part of them
https://github.com/facebookincubator/velox/blob/main/velox/type/Conversions.h#L231

File: velox/type/DecimalUtil.h [hunk 7]
```diff
@@ -147,6 +147,36 @@ class DecimalUtil {
...
 
+  template <typename TInput, typename TOutput>
+  inline static std::optional<TOutput> rescaleFloating(
+      const TInput inputValue,
+      const int toPrecision,
+      const int toScale) {
+    if (!std::isfinite(inputValue)) {
+      return std::nullopt;
+    }
+    auto toValue =
```
Comment (jinchengchenghh) [inline velox/type/DecimalUtil.h:158]: Here may overflow, please test the double::max

Comment (wuxueyang96) [reply velox/type/DecimalUtil.h:158]: Yes, it will overflow here, so I have checked below.

Comment (jinchengchenghh) [reply velox/type/DecimalUtil.h:158]: if the inputValue is double::max, you check the unexpected overflowd double value, not the value after multiply.

Comment (wuxueyang96) [reply velox/type/DecimalUtil.h:158]: If the value after multiplication is overflowed, the 'toValue' is std::inf or something like this. It will not pass the check below.
But indeed we could set the default value of 'overflow' by checking if 'toValue' is overflowed.

done.

**Comment by wuxueyang96** [comment]
@jinchengchenghh Could you take another look?

File: velox/type/DecimalUtil.h [hunk 8]
```diff
@@ -147,6 +147,36 @@ class DecimalUtil {
...
+        inputValue * static_cast<TInput>(DecimalUtil::kPowersOfTen[toScale]);
+    bool isOverflow = !std::isfinite(toValue);
+    toValue = toValue - std::floor(toValue) >= 0.5 ? std::ceil(toValue)
```
Comment (jinchengchenghh) [inline velox/type/DecimalUtil.h:161]: Can we use exists function to convert double to int128 or int64? https://github.com/facebookincubator/velox/blob/main/velox/type/Conversions.h#L231

Comment (wuxueyang96) [reply velox/type/DecimalUtil.h]: I don't think this existing function works in this case. This function has no unit test and we know nothing about this function.

If we use `velox::util::Converter<TypeKind::HUGEINT>::cast((double) kLongDecimalMax)`, it will throw exception like 'Loss of precision during arithmetic conversion'. If we use `velox::util::Converter<TypeKind::HUGEINT, void, true>::cast((double)kLongDecimalMax)`, the return value is definitely wrong.

Could you give me an example about using this function to convert double to int128? @jinchengchenghh

Comment (jinchengchenghh) [reply velox/type/DecimalUtil.h:161]: It is base function, now it may be used nowhere.
If it has error, we could ask the author for help. Can you help look into this issue? @majetideepak

Comment (wuxueyang96) [reply velox/type/DecimalUtil.h]: @jinchengchenghh I read the source code of 'velox::util::Converter', refer to https://github.com/facebook/folly/blob/main/folly/Conv.h#L1345, this function is suitable here except it will throw an exception if a double number exceeds the limit of an integral number. I think we could catch this exception here and judge if this casting is overflowed. But I am not sure. Could you give me a favor to check if we could use this function in this way?

Comment (jinchengchenghh) [reply velox/type/DecimalUtil.h]: I think it's OK, and you should use the `truncate` from config https://github.com/facebookincubator/velox/blob/main/velox/expression/CastExpr-inl.h#L255

File: velox/type/DecimalUtil.h [hunk 9]
```diff
@@ -147,6 +148,44 @@ class DecimalUtil {
...
+    bool isOverflow = !std::isfinite(toValue);
+    if (toValue > static_cast<TInput>(std::numeric_limits<TOutput>::max()) ||
+        toValue < static_cast<TInput>(std::numeric_limits<TOutput>::min())) {
+      isOverflow = true;
+    }
+
+    TOutput rescaledValue;
+    try {
+      rescaledValue = static_cast<TOutput>(
+          util::Converter<TypeKind::HUGEINT>::cast(toValue));
```
Comment (jinchengchenghh) [inline velox/type/DecimalUtil.h:171]: It could be HUGEINT for long decimal and BIGINT for short decimal.

File: velox/type/DecimalUtil.h [hunk 10]
```diff
@@ -147,6 +148,44 @@ class DecimalUtil {
...
+    } catch (const std::exception& e) {
+      LOG(ERROR) << "Failed to cast " << toValue << " to hugeInt";
```
Comment (jinchengchenghh) [inline velox/type/DecimalUtil.h:173]: nt: hugeInt

**Comment by jinchengchenghh** [comment]
Please remove `.` from title

File: velox/type/DecimalUtil.h (hunk 8 above)
Comment (wuxueyang96) [reply velox/type/DecimalUtil.h]: No, we can't. Look at https://github.com/facebookincubator/velox/blob/main/velox/type/Conversions.h#L236, if we use the truncate, and the return value's type is int128_t, it will return `std::numeric_limits<int128_t>::max()` directly. That's the reason why I think If we use `velox::util::Converter<TypeKind::HUGEINT, void, true>::cast((double)kLongDecimalMax)`, the return value is definitely wrong.

Besides, this config is used to cast a floating number to an integral number, but we want to cast a floating number to a decimal number. Although we need to cast a floating number to an integral number here, we don't want to get a truncated value here because we need to know whether the overflow occurs here.

**Comment by jinchengchenghh** [comment]
What if the double value is NaN?

**Comment by wuxueyang96** [comment]
> What if the double value is NaN?

It will return null. At the beginning of `rescaleFloating`, https://github.com/facebookincubator/velox/pull/5767/files#diff-c6d2f99b68772f908f2f3e8b979b3c2894f790f20af68684e7b0400bed969389R156

**Comment by jinchengchenghh** [comment]
> > What if the double value is NaN?
> 
> It will return null. At the beginning of `rescaleFloating`, https://github.com/facebookincubator/velox/pull/5767/files#diff-c6d2f99b68772f908f2f3e8b979b3c2894f790f20af68684e7b0400bed969389R156

It should fail, I think. https://github.com/facebookincubator/velox/blob/main/velox/type/Conversions.h#L249

**Comment by wuxueyang96** [comment]
> > > What if the double value is NaN?
> > 
> > 
> > It will return null. At the beginning of `rescaleFloating`, https://github.com/facebookincubator/velox/pull/5767/files#diff-c6d2f99b68772f908f2f3e8b979b3c2894f790f20af68684e7b0400bed969389R156
> 
> It should fail, I think. https://github.com/facebookincubator/velox/blob/main/velox/type/Conversions.h#L249

done. And should it fail when the value is INFINITY?

**Comment by jinchengchenghh** [comment]
Yes, but Velox may not fully support INFINITY.

**Comment by jinchengchenghh** [comment]
One last question, if `overflow`, should we fail at once?
 @rui-mo Could you help review this PR? It will be used in TPCDS.

**Comment by wuxueyang96** [comment]
> One last question, if `overflow`, should we fail at once? @rui-mo Could you help review this PR? It will be used in TPCDS.

For fast-fail, it's better to fail once there is `overflow`. But I think there will hardly see a `overflow` in the common situation, and failing at once if there is `overflow` will add some 'if-else', that maybe increases the execution time. So just for me, I think it's better to check at the final.

File: velox/type/DecimalUtil.h [hunk 11]
```diff
@@ -147,6 +148,43 @@ class DecimalUtil {
     return static_cast<TOutput>(rescaledValue);
   }
 
+  template <typename TInput, typename TOutput>
+  inline static std::optional<TOutput> rescaleFloating(
```
Comment (rui-mo) [inline velox/type/DecimalUtil.h:152]: Can we add some comments for this function?

File: velox/type/DecimalUtil.h [hunk 12]
```diff
@@ -147,6 +148,43 @@ class DecimalUtil {
...
+      const TInput inputValue,
+      const int toPrecision,
+      const int toScale) {
+    auto toValue =
+        inputValue * static_cast<TInput>(DecimalUtil::kPowersOfTen[toScale]);
```
Comment (rui-mo) [inline velox/type/DecimalUtil.h:157]: If I understand correctly, `toValue` can not be larger than `static_cast<TInput>(std::numeric_limits<TOutput>::max())`, because the returned value is of `TInput` type. Same for negative case. Is below check still needed?

```
    if (toValue > static_cast<TInput>(std::numeric_limits<TOutput>::max()) ||
        toValue < static_cast<TInput>(std::numeric_limits<TOutput>::min())) {
      isOverflow = true;
    }
```

File: velox/type/DecimalUtil.h [hunk 13]
```diff
@@ -147,6 +148,43 @@ class DecimalUtil {
...
+
+    bool isOverflow = !std::isfinite(toValue);
```
Comment (rui-mo) [inline velox/type/DecimalUtil.h:159]: If `toValue` is not infinitty, `isOverflow` is set as true. Is this correct?

File: velox/expression/CastExpr.cpp [hunk 14]
```diff
@@ -439,6 +439,10 @@ VectorPtr CastExpr::applyDecimal(
       applyIntToDecimalCastKernel<int32_t, toDecimalType>(
           rows, input, context, toType, castResult);
       break;
+    case TypeKind::DOUBLE:
+      applyDoubleToDecimal<double, toDecimalType>(
```
Comment (rui-mo) [inline velox/expression/CastExpr.cpp:443]: We may also need to document this casting behavior in https://github.com/facebookincubator/velox/blob/main/velox/docs/functions/presto/conversion.rst.

File: velox/type/DecimalUtil.h (hunk 12 above)
Comment (wuxueyang96) [reply velox/type/DecimalUtil.h:157]: No, `toValue` could be larger than `static_cast<TInput>(std::numeric_limits<TOutput>::max())`, because TInput may be double, and TOutput may be int64_t. It is very usual that a double value is bigger than int64_t.
For example, if `inputValue` is 9223372036854775808, `toScale` is 2 and TOutput is int64_t, then `toValue` should be 922337203685477580800 in double type, it is bigger than `static_cast<double>(std::numeric_limits<int64_t>::max())`'value 9223372036854775807.

File: velox/type/DecimalUtil.h (hunk 13 above)
Comment (wuxueyang96) [reply velox/type/DecimalUtil.h:159]: The function `std::isfinite` will return false only if the `toValue` is `NAN` or `INFINITY`. So if `toValue` is not infinity or nan, `isOverflow` will be set as false.

File: velox/type/DecimalUtil.h (hunk 12 above)
Comment (rui-mo) [reply velox/type/DecimalUtil.h:157]: Thanks for the details. If TOutput is int128_t, do we still need this check? Because double value cannot be larger the maximum of int128_t. Maybe this part can be moved to below else branch for bigint handling.

File: velox/type/DecimalUtil.h (hunk 13 above)
Comment (rui-mo) [reply velox/type/DecimalUtil.h:159]: Sorry, a missing reading to `std::isinf`.

File: velox/expression/tests/CastExprTest.cpp [hunk 15]
```diff
@@ -943,6 +1063,10 @@ TEST_F(CastExprTest, toString) {
   ASSERT_EQ("cast((a) as ARRAY<VARCHAR>)", exprSet.exprs()[1]->toString());
 }
 
+TEST_F(CastExprTest, floatingToDecimal) {
+  testFloatingToDecimalCasts<double>();
```
Comment (rui-mo) [inline velox/expression/tests/CastExprTest.cpp:1067]: Can we test for float type as well?

File: velox/expression/CastExpr.cpp [hunk 16]
```diff
@@ -439,6 +439,10 @@ VectorPtr CastExpr::applyDecimal(
       applyIntToDecimalCastKernel<int32_t, toDecimalType>(
           rows, input, context, toType, castResult);
       break;
+    case TypeKind::DOUBLE:
```
Comment (rui-mo) [inline velox/expression/CastExpr.cpp:442]: Do we need to support float type as well? Looks the logics are similar.

File: velox/expression/CastExpr.h [hunk 17]
```diff
@@ -211,6 +211,17 @@ class CastExpr : public SpecialForm {
       const TypePtr& toType,
       VectorPtr& castResult);
 
+  template <
+      typename TInput,
+      typename TOutput,
+      typename std::enable_if_t<std::is_floating_point_v<TInput>, bool> = true>
```
Comment (rui-mo) [inline velox/expression/CastExpr.h:217]: std::is_floating_point_v should returns true for both float and double type.

File: velox/expression/CastExpr.h [hunk 18]
```diff
@@ -211,6 +211,17 @@ class CastExpr : public SpecialForm {
...
+  void applyDoubleToDecimal(
```
Comment (rui-mo) [inline velox/expression/CastExpr.h:218]: Maybe rename to applyFloatingToDecimal?

File: velox/expression/tests/CastExprTest.cpp (hunk 15 above)
Comment (wuxueyang96) [reply velox/expression/tests/CastExprTest.cpp:1067]: Currently, no, we can't. The float number has a very low precision. For example:

```cpp
float num = 99999.99f;
std::cout << std::fixed << num << std::endl;
```

The output will be `99999.992188`.

In presto, the value about `select cast(99999.99 as decimal(10,4));` is `99999.9900`, but in this unit test, the value is `99999.9921`.

So in fact, we haven't support float yet.

File: velox/expression/tests/CastExprTest.cpp [hunk 19]
```diff
@@ -943,6 +943,133 @@ TEST_F(CastExprTest, toString) {
   ASSERT_EQ("cast((a) as ARRAY<VARCHAR>)", exprSet.exprs()[1]->toString());
 }
 
+TEST_F(CastExprTest, doubleToDecimal) {
+  // floating to short decimal
```
Comment (rui-mo) [inline velox/expression/tests/CastExprTest.cpp:947]: Nit: please follow the comment style by capitalize the first letter and add `.` at the end.

File: velox/expression/tests/CastExprTest.cpp [hunk 20]
```diff
@@ -943,6 +943,133 @@ TEST_F(CastExprTest, toString) {
...
+  auto input =
+      makeFlatVector<double>({-3333.03, -2222.02, -1.0, 0.00, 100, 99999.99});
+  testComplexCast(
+      "c0",
+      input,
+      makeFlatVector<int64_t>(
+          {-33'330'300, -22'220'200, -10'000, 0, 1'000'000, 999'999'900},
+          DECIMAL(10, 4)));
+
+  // floating to long decimal
```
Comment (rui-mo) [inline velox/expression/tests/CastExprTest.cpp:957]: Same.

File: velox/expression/tests/CastExprTest.cpp [hunk 21]
```diff
@@ -943,6 +943,133 @@ TEST_F(CastExprTest, toString) {
...
+  auto numberBiggerThanInt64Max = static_cast<double>(
+      static_cast<int128_t>(std::numeric_limits<int64_t>::max()) + 1);
+  auto numberSmallerThanInt64Min = static_cast<double>(
+      static_cast<int128_t>(std::numeric_limits<int64_t>::min()) - 1);
+  auto numberBiggerThanDecimal20 =
+      static_cast<double>(DecimalUtil::kLongDecimalMax);
+  auto numberSmallerThanDecimal20 =
+      static_cast<double>(DecimalUtil::kLongDecimalMin);
+
+  // Expected failures
```
Comment (rui-mo) [inline velox/expression/tests/CastExprTest.cpp:984]: Same.

File: velox/type/DecimalUtil.h [hunk 22]
```diff
@@ -147,6 +148,50 @@ class DecimalUtil {
...
+        inputValue * static_cast<double>(DecimalUtil::kPowersOfTen[toScale]);
+
+    bool isOverflow = !std::isfinite(toValue);
+    TOutput rescaledValue;
+    try {
+      if constexpr (std::is_same<TOutput, int128_t>::value) {
+        rescaledValue = util::Converter<TypeKind::HUGEINT>::cast(toValue);
+      } else {
+        rescaledValue = util::Converter<TypeKind::BIGINT>::cast(toValue);
+      }
```
Comment (rui-mo) [inline velox/type/DecimalUtil.h:178]: Would below work?
```
rescaledValue = util::Converter<CppToType<TOutput>::typeKind>::cast(toValue);
```

**Review by rui-mo — COMMENTED** [review]
Thanks for working on this. Just several nits.

File: velox/type/DecimalUtil.h [hunk 23]
```diff
@@ -147,6 +148,50 @@ class DecimalUtil {
     return static_cast<TOutput>(rescaledValue);
   }
 
+  /**
+   * Rescale a double value to decimal value.
+   *
+   * Use `folly::convertTo` to convert a double value to int128_t or int64_t, it
```
Comment (jinchengchenghh) [inline velox/type/DecimalUtil.h:154]: Please use /// for comments in header file.
https://github.com/facebookincubator/velox/blob/main/CODING_STYLE.md#comments

**Comment by wuxueyang96** [comment]
@jinchengchenghh @rui-mo PTAL

**Comment by rui-mo** [comment]
@mbasmanova @Yuhta Could you help take a review? Thanks!

File: velox/docs/functions/presto/conversion.rst [hunk 24]
```diff
@@ -501,6 +501,27 @@ Invalid examples
...
+Casting a double precision floating-point type number to a decimal of given precision
+and scale is allowed if the input value can be represented by the precision and scale.
+Casting from invalid input values throws.
+
+Valid example
+
+::
+
+  SELECT cast(0.12 as decimal(4,4)); -- decimal '0.1200'
+  SELECT cast(0.12 as decimal(4,1)); -- decimal '0.1'
```
Comment (mbasmanova) [inline velox/docs/functions/presto/conversion.rst:516]: Can you document truncating / rounding behavior?

```
presto:di> SELECT cast(0.19 as decimal(4,1));
 _col0
-------
 0.2
(1 row)
```

File: velox/expression/CastExpr-inl.h [hunk 25]
```diff
@@ -206,6 +206,35 @@ void CastExpr::applyIntToDecimalCastKernel(
       });
 }
 
+template <typename TOutput>
```
Comment (mbasmanova) [inline velox/expression/CastExpr-inl.h:209]: Is this template parameter needed? Isn't it always int128_t?

File: velox/expression/CastExpr-inl.h [hunk 26]
```diff
@@ -206,6 +206,35 @@ void CastExpr::applyIntToDecimalCastKernel(
...
+void CastExpr::applyDoubleToDecimal(
+    const SelectivityVector& rows,
+    const BaseVector& input,
+    exec::EvalCtx& context,
+    const TypePtr& toType,
+    VectorPtr& castResult) {
+  auto sourceVector = input.as<SimpleVector<double>>();
+  auto castResultRawBuffer =
```
Comment (mbasmanova) [inline velox/expression/CastExpr-inl.h:217]: naming: rawResults

File: velox/expression/CastExpr-inl.h (hunk 25 above)
Comment (wuxueyang96) [reply velox/expression/CastExpr-inl.h:209]: It could be int64_t.

File: velox/expression/CastExpr-inl.h [hunk 27]
```diff
@@ -206,6 +206,35 @@ void CastExpr::applyIntToDecimalCastKernel(
...
+      castResult->asUnchecked<FlatVector<TOutput>>()->mutableRawValues();
+  const auto& toPrecisionScale = getDecimalPrecisionScale(*toType);
```
Comment (mbasmanova) [inline velox/expression/CastExpr-inl.h:219]: use 'const auto'; getDecimalPrecisionScale returns a temp variable; can't store a reference to it

File: velox/expression/CastExpr-inl.h [hunk 28]
```diff
@@ -206,6 +206,35 @@ void CastExpr::applyIntToDecimalCastKernel(
...
+          return;
+        }
+        auto rescaledValue = DecimalUtil::rescaleDouble<TOutput>(
+            sourceVector->valueAt(row),
+            toPrecisionScale.first,
+            toPrecisionScale.second);
+        if (rescaledValue.has_value()) {
+          castResultRawBuffer[row] = rescaledValue.value();
+        } else {
+          castResult->setNull(row, true);
```
Comment (mbasmanova) [inline velox/expression/CastExpr-inl.h:233]: Is this the right behavior? Shouldn't this code path raise an error?

File: velox/type/DecimalUtil.h [hunk 29]
```diff
@@ -147,6 +148,45 @@ class DecimalUtil {
...
+  /// it will throw an exception like 'loss of precision' when the `toValue`
+  /// overflows the limits of int128_t or int64_t. The exception will be caught
+  /// here to judge whether there is a overflow.
+  ///
+  /// Normally, return the rescaled value. Otherwise, if the `toValue` overflows
+  /// the TOutput's limits or the `toValue` exceeds the precision's limits, it
+  /// will throw an exception.
+  template <typename TOutput>
+  inline static std::optional<TOutput> rescaleDouble(
+      const double inputValue,
```
Comment (mbasmanova) [inline velox/type/DecimalUtil.h:163]: drop 'const'

File: velox/type/DecimalUtil.h [hunk 30]
```diff
@@ -147,6 +148,45 @@ class DecimalUtil {
...
+  /// Use `folly::convertTo` to convert a double value to int128_t or int64_t,
+  /// it will throw an exception like 'loss of precision' when the `toValue`
+  /// overflows the limits of int128_t or int64_t. The exception will be caught
+  /// here to judge whether there is a overflow.
+  ///
+  /// Normally, return the rescaled value. Otherwise, if the `toValue` overflows
+  /// the TOutput's limits or the `toValue` exceeds the precision's limits, it
+  /// will throw an exception.
+  template <typename TOutput>
+  inline static std::optional<TOutput> rescaleDouble(
```
Comment (mbasmanova) [inline velox/type/DecimalUtil.h:162]: Why is the output an optional? When do you return unset optional?

It would be nice to extract this utility into a separate PR and add tests. CC: @majetideepak

File: velox/type/DecimalUtil.h [hunk 31]
```diff
@@ -147,6 +148,45 @@ class DecimalUtil {
...
+      const int toScale) {
+    auto toValue =
+        inputValue * static_cast<double>(DecimalUtil::kPowersOfTen[toScale]);
+
+    bool isOverflow = !std::isfinite(toValue);
+    TOutput rescaledValue;
+    try {
+      rescaledValue =
+          util::Converter<CppToType<TOutput>::typeKind>::cast(toValue);
+    } catch (const std::exception& e) {
```
Comment (mbasmanova) [inline velox/type/DecimalUtil.h:174]: Exceptions are super expensive and make try_cast very inefficient. Please, find a better way. See https://github.com/facebookincubator/velox/pull/5801/

File: velox/expression/CastExpr-inl.h (hunk 25 above)
Comment (mbasmanova) [reply velox/expression/CastExpr-inl.h:209]: Got it. Thank you for clarifying. It would be nice to add a comment explaining that.

File: velox/expression/tests/CastExprTest.cpp [hunk 32]
```diff
@@ -943,6 +943,133 @@ TEST_F(CastExprTest, toString) {
...
+      static_cast<double>(DecimalUtil::kLongDecimalMax);
+  auto numberSmallerThanDecimal20 =
+      static_cast<double>(DecimalUtil::kLongDecimalMin);
+
+  // Expected failures.
+  VELOX_ASSERT_THROW(
+      testComplexCast(
+          "c0",
+          makeFlatVector<double>(
+              std::vector<double>{9999999999999999999999.99}),
```
[the remark this answers is not in the document]
Comment (jinchengchenghh) [reply velox/expression/tests/CastExprTest.cpp:989]: If only 1 value, please use makeConstant<double>(99.99, 1);

File: velox/expression/tests/CastExprTest.cpp [hunk 33]
```diff
@@ -943,6 +943,133 @@ TEST_F(CastExprTest, toString) {
...
+          makeFlatVector<int64_t>(std::vector<int64_t>{0}, DECIMAL(10, 2))),
```
[the remark this answers is not in the document]
Comment (jinchengchenghh) [reply velox/expression/tests/CastExprTest.cpp:990]: `makeConstant<int64_t>(0, 1, DECIMAL(10, 2));`

**Comment by wuxueyang96** [comment]
@karteekmurthys Could you review #6051 before this pr? #6051 should be merged before this pr. Thanks.

**Comment by wuxueyang96** [comment]
@mbasmanova Please take a look, thanks!

**Review by mbasmanova — COMMENTED** [review]
@majetideepak Deepak, would you help take a first pass?

**Comment by rui-mo** [comment]
wuxueyang96 This PR is marked as stale, could you help rebase it? Thanks!

**Comment by yzhang1991** [comment]
Hi @majetideepak, we need this PR for the IBM BI workload, can you help review?

**Review by karteekmurthys — APPROVED** [review]
@wuxueyang96 Thanks for your patience and adding this change.
LGTM! please fix the CI test failures. The messaging has 1e+22 vs 1E22.

```
204: ../../velox/expression/tests/CastExprTest.cpp:1534: Failure
204: Value of: e.message().find(fmt::format( "Failed to cast from DOUBLE to {}: {}. {}", type->toString(), value, detail)) != std::string::npos
204:   Actual: false
204: Expected: true
204: Expected error message to contain 'Failed to cast from DOUBLE to DECIMAL(10,2): 1e+22. Rescaled value is overflowed.', but received 'Failed to cast from DOUBLE to DECIMAL(10,2): 1E22. Rescaled value is overflowed.'.
204: ../../velox/expression/tests/CastExprTest.cpp:1534: Failure
204: Value of: e.message().find(fmt::format( "Failed to cast from DOUBLE to {}: {}. {}", type->toString(), value, detail)) != std::string::npos
204:   Actual: false
204: Expected: true
204: Expected error message to contain 'Failed to cast from DOUBLE to DECIMAL(10,2): 9.223372036854776e+18. Rescaled value is overflowed.', but received 'Failed to cast from DOUBLE to DECIMAL(10,2): 9223372036854776000. Rescaled value is overflowed.'.
204: ../../velox/expression/tests/CastExprTest.cpp:1534: Failure
```
